src/bench_alkahest.rs

```diff
@@ -0,0 +1,46 @@
+use criterion::{black_box, Criterion};
+use alkahest::{
+    Pack, Schema, Unpacked,
+};
+
+pub fn bench<T, D>(name: &'static str, c: &mut Criterion, data: D, read: fn(Unpacked<'_, T>))
+where
+    T: Schema,
+    D: Pack<T> + Copy,
+{
+    let mut group = c.benchmark_group(format!("{}/alkahest", name));
+
+    // This leak is required because of ICE in rustc if `R` has bound like this `for<'a> Fn(Unpacked<'a, T>)`.
+    let mut buffer = vec![0u64; 1 << 1 << 22];
+    let bytes: &mut [u8] = bytemuck::cast_slice_mut(&mut buffer[..]);
+    let mut size = 0;
+
+    group.bench_function("serialize", |b| {
+        b.iter(|| {
+            black_box(
+                size = alkahest::write::<T, _>(bytes, black_box(data))
+            );
+        })
+    });
+
+    group.bench_function("access (validated)", |b| {
```

Comment:
This benchmark appears to be unvalidated. The docs for `alkahest::read` state that the function will panic if the data does not conform to the expected schema.

Reply:
I guess I don't have clean understanding for what is "unvalideted" and "validated" mean then.
 I was under impression it's like "unsafe" and "safe".

Reply:
(David correct me if I'm wrong): I believe "validated" means that you have checked that the deserialized values make sense. e.g. for an enum, if you have the following variants:

```rust
enum Example {
    Foo,
    Bar,
    Baz,
}
```

and encode them as `0`, `1`, and `2`, if you deserialize a message with an enum discriminant of `250`, that would be bad, as it's an "impossible value".

Reply:
"Bad" can mean many things.
If deserializing `250` would lead to UB this is bad-bad.
Alkahest would just panic. IF it would read the value, not if bad value just appear somewhere in nested branching structure and the reading code never touches that branch.

Reply:
Hmm, you raise a good point. This is always a bit thorny for terminology. I think that validated definitely implies no undefined behavior on invalid data, which a panic satisfies. The question is whether it should further imply that it doesn't panic.

Perhaps we could make it clearer with different language? Flatbuffers uses "verified" vs "unverified" because that's how the developers refer to it. Maybe something like "checked access"?

Reply:
What can happen in case of invalid data?

* Error
* Error during while traversing
* Panic
* Panic during while traversing
* UB

Can we define 5 terms and explain them before result tables.

Reply:
Sure, I think that would work. I think the categories would be:

- Upfront, error (flatbuffers/rkyv)
- Upfront, panic
- On-demand, error (capnproto)
- On-demand, panic (alkahest)
- Unvalidated

Then if a framework supports multiple modes (e.g. upfront error and unvalidated) then there will be separate benches for each.